File: app/src/main/java/com/awscherb/cardkeeper/data/handler/BaseHandler.java
package com.awscherb.cardkeeper.data.handler;

import com.awscherb.cardkeeper.data.model.BaseModel;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;


abstract class BaseHandler<T extends BaseModel> {

    protected static final String FIELD_ID = "id";

    //================================================================================
    // Get
    //================================================================================

    abstract Single<T> getObject(long id);

    abstract Observable<List<T>> listObjects();

    //================================================================================
    // Create
    //================================================================================

    abstract Single<T> createObject(T object);

    //================================================================================
    // Update
    //================================================================================

    abstract Single<T> updateObject(T object);

    //================================================================================
    // Delete
    //================================================================================

    abstract Completable deleteObject(T object);

}

File: app/src/main/java/com/awscherb/cardkeeper/data/handler/BaseHandler.kt
package com.awscherb.cardkeeper.data.handler

import io.reactivex.CompletableTransformer
import io.reactivex.FlowableTransformer
import io.reactivex.SingleTransformer
import io.reactivex.schedulers.Schedulers

abstract class BaseHandler {
    fun <T> scheduleSingle() = SingleTransformer<T, T> { s -> s.subscribeOn(Schedulers.io()) }
    fun scheduleCompletable() = CompletableTransformer { s -> s.subscribeOn(Schedulers.io()) }
    fun <T> scheduleFlowable() = FlowableTransformer<T,T> { s -> s.subscribeOn(Schedulers.io()) }
}